File: grind75/week2/021-Find-All-Anagrams-in-a-String-Medium/021-Find-All-Anagrams-in-a-String-Medium.cpp
// link: https://leetcode.com/problems/find-all-anagrams-in-a-string/description/
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        // cannot make anagrams if len of p is greater than the len of s
        if (s.length() < p.length())
            return {};
        
        unordered_map<char, int> pMap;
        unordered_map<char, int> sMap;
        vector<int> ans;

        for (char c : p)
            pMap[c]++;
        
        for(int i = 0; i < s.length(); i++)
        {
            sMap[s[i]]++;

            if(i >= p.length())
            {
                char leftChar = s[i - p.length()];
                if(sMap[leftChar] == 1)
                    sMap.erase(leftChar);
                else
                    sMap[leftChar]--;
            }

            if(pMap == sMap)
                ans.push_back(i - p.length() + 1);
                 
        }


        return ans;
    }
};

File: grind75/week2/021-Find-All-Anagrams-in-a-String-Medium/021-Find-All-Anagrams-in-a-String-Medium.py
# link: https://leetcode.com/problems/find-all-anagrams-in-a-string/description/

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        # cannot make anagrams if len of p is greater than len of s
        if len(s) < len(p):
            return []

        p_map = collections.Counter(p)
        s_map = {}
        ans = []
        for i in range(len(s)):
            if s[i] not in  s_map:
                s_map[s[i]] = 1
            else:
                s_map[s[i]] += 1
            
            if i >= len(p):
                if s_map[s[i - len(p)]] == 1:
                    del s_map[s[i - len(p)]]
                else:
                    s_map[s[i - len(p)]] -= 1

            if p_map == s_map:
                ans.append(i - len(p) + 1)

        return ans

        
